docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+Decorators belong to the realm of somewhat less-known Python features,
+but nevertheless, they are very simple:
+
+    >>> def application(wrapped_method):
+    >>>     # A decorator takes a function and returns a function.
+    >>>     # Note, that a method is just a function, and `self` is just an argument.
+    >>>     # The returned function is what you actually use later. 
+    >>>     # It typically makes a number of calls of the wrapped function, 
+    >>>     # in addition doing some other stuff, pre- and post- processing 
+    >>>     # in our case.
+    >>>     def returned_function(self, *args, **kwargs):
+    >>>         annotated_args = []
+    >>>         for arg in args:
+    >>>             if isinstance(arg, theano.Variable):
+    >>>                 new_arg = arg.copy()
+    >>>                 new_arg.tag.stone = self
+    >>>                 new_arg.tag.role = 'input'
+    >>>                 # In addition to `.tag` attribute, we also set the `.name` attribute.
+    >>>                 # It is recognized by Theano debugging tools.
+    >>>                 new_arg.name = '{}_{}'.format(self.__class__.__name__, 'input')
+    >>>                 annotated_args.append(new_arg)
+    >>>             else:
+    >>>                 annotated_args.append(arg)
+    >>>         annotated_kwargs = {}
+    >>>         for key, value in kwargs.items():
+    >>>             if instance(value, theano.Variable):
+    >>>                 new_value = value.copy()
+    >>>                 new_value.tag.stone = self
+    >>>                 new_value.tag.role = 'input'
+    >>>                 new_value.name = '{}_{}'.format(self.__class__.__name__, 'input')
+    >>>                 annotated_kwargs[key] = new_value
+    >>>             else:
+    >>>                 annotated_kwargs[key] = value
+    >>>         output = wrapped_method(self, *annotated_args, **annotated_kwargs).copy()
+    >>>         output.tag.stone = self
+    >>>         output.tag.role = 'output'
+    >>>         output.name = '{}_{}'.format(self.__class__.__name__, 'output')
+    >>>         return output
+    >>>     return returned_function
+    >>>     
+    >>>     
+    >>> # Now Tanh becomes simple
+    >>> class Tanh(Stone):
+    >>>     @application
+    >>>     def apply(self, x):
+    >>>         return tensor.tanh(x)
+    >>>     
+    >>> class Softmax(Stone):
+    >>>     @application
+    >>>     def apply(self, x):
+    >>>         return tensor.nnet.softmax(x)
+    >>>     
+    >>> x = tensor.matrix('x')
+    >>> x.tag.test_value = numpy.zeros((10, 20))
+    >>> theano.printing.debugprint(Tanh().apply(x))
+    Elemwise{identity} [@A] 'Tanh_output'   
+     |Elemwise{tanh,no_inplace} [@B] ''   
+       |Elemwise{identity} [@C] 'Tanh_input'   
+         |x [@D]
+
+
+Now, ``MultiLayer`` can annotate its inputs and outputs usign the same
+decorator! Note, that now activations have a different interface: they are
```

Comment:
I would flip the sentence in "Now, we can annotate the inputs and outputs of ``MultiLayer`` using the same `@application` decorator!"